### PR TITLE
[Gemma] Fix eager attention

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Fixes the Gemma "eager" attention implementation, which is the default for torch versions <= 2.1. This issue was reported on the [Hub discussions](https://huggingface.co/google/gemma-7b-it/discussions/13) and by @osanseviero from the model cards/blog post. 

The PR also includes a set of slow tests to ensure:
1. Logits equivalence between {eager, sdpa, flash attention 2}
2. Expected generation results for {eager, sdpa, flash attention 2}

=> these tests confirm all attention implementations work and have equivalence between back-ends

### CODE REVIEW

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -276,7 +276,7 @@ def forward(
 
         attn_output = attn_output.transpose(1, 2).contiguous()
 
-        attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)
+        attn_output = attn_output.view(bsz, q_len, -1)
```
Comment (sanchit-gandhi) [inline src/transformers/models/gemma/modeling_gemma.py:279]: This is the only modelling code change required - the remainder of the changes in this PR are logit + integration tests

**Comment by sanchit-gandhi** [comment]
Feel free to merge at your convenience @ArthurZucker

**Review by ArthurZucker — APPROVED** [review]

**Comment by ArthurZucker** [comment]
Thanks for the catch. A real pity our 280 prior test did not catch this! 🤗

**Comment by sanchit-gandhi** [comment]
I'll look at integrating 2 tests in `test_modeling_common.py` that can automatically test for eager vs sdpa and eager vs fa2. Currently, this is done on an ad-hoc basis for each model (e.g. Whisper [here](https://github.com/huggingface/transformers/blob/2a9b1f80c45cab19b542bc7cc004937d39d6f6fb/tests/models/whisper/test_modeling_whisper.py#L891))

**Comment by ArthurZucker** [comment]
It's also that it tests dummy models, but here the head_dim != hidden / head! So a case no studied

**Comment by fxmarty** [comment]
@sanchit-gandhi Not true, see https://github.com/huggingface/transformers/blob/2cc8cf6ce7ae0416561acbb639df4bbc5f409b6f/tests/test_modeling_common.py#L3431. It is just a slow test.

**Comment by ArthurZucker** [comment]
eager vs sdpa passed 😉
